src/sql/src/query_model/attribute/relation_type.rs

```diff
@@ -0,0 +1,511 @@
+
+        // the type of All and Existential quantifiers is always BOOLEAN NOT NULL
+        if matches!(
+            quantifier.quantifier_type,
+            QuantifierType::Existential | QuantifierType::All
+        ) {
+            return ColumnType {
+                scalar_type: ScalarType::Bool,
+                nullable: false,
+            };
+        }
+
+        let input_box = quantifier.input_box();
+        let column_type = &input_box.attributes.get::<RelationType>()[cref.position];
+
+        let parent_box = self.get_box(quantifier.parent_box);
+        let nullable = if enclosing_box_id == parent_box.id {
+            if quantifier.quantifier_type == QuantifierType::Scalar {
+                // scalar quantifiers can always be NULL if the subquery is empty
+                true
+            } else {
+                use BoxType::*;
+                // In general, nullability is influenced by the parent box of the quantifier.
+                match &parent_box.box_type {
+                    // The type can be restricted to NOT NULL if it is used in a Select
+                    // box with a predicate that rejects nulls in that ColumnReference.
+                    Select(..) if select_rejects_nulls(&parent_box, cref) => false,
+                    // The type should be widened to NULL if it is used in an OuterJoin
+                    // box where the opposite side is preserving.
+                    OuterJoin(..) if outer_join_introduces_nulls(&parent_box, cref) => true,
+                    // The type of aggregates in grouping boxes without keys (a.k.a.
+                    // global aggregates) can always be NULL. This works correctly only
+                    // under the assumption that the output columns of a Grouping box
+                    // are always trivial ColumnReference expressions.
+                    Grouping(grouping) if grouping.key.is_empty() => true,
+                    // The type should be widened to NULL if it is used in a Union
+                    // box with at least one quantifier in the same position being NULL.
+                    Union if union_introduces_nulls(&parent_box, cref) => true,
+                    // The type should be restricted to NOT NULL if it is used in an
+                    // Intersect box with at least one quantifier in the same position
+                    // being NOT NULL.
+                    Intersect if intersect_rejects_nulls(&parent_box, cref) => false,
+                    // otherwise, just inherit the nullable information from the input
+                    _ => column_type.nullable.clone(),
+                }
+            }
+        } else {
+            // If the enclosing box is different from the parent box (which
+            // can happen in a graph derived from a correlated subquery), the
+            // RelationType of the latter might not be fully derived. In this
+            // case, we therefore just inherit the nullability of the input.
```

Comment:
I'm trying to understand why is it that the `RelationType` of the input box of the quantifier has always been derived but the `RelationType` of the parent box of the quantifier might not be fully derived.

Is it that we relying on the following two conditions?
1) the correlated subquery is always to the right of the outer query
2) we will always visit left quantifiers before right quantifiers.

If this is the case, we should put down somewhere (possibly rewrite/mod.rs) that rewrites that run before decorrelation cannot change the order of the quantifiers in a box.

Reply:
> I'm trying to understand why is it that the RelationType of the input box of the quantifier has always been derived but the RelationType of the parent box of the quantifier might not be fully derived.

This is because derivation proceeds in a bottom-up manner (post-visit), so the `RelationType` of all children will be derived before the `RelationType` of the parent.

So yes, we are relying on the two conditions that you mention. A `Model` instance that has a quantifier `Qi` that depends on `Qj` where `Qi` appears before `Qj` is invalid. We should rather have a `Validator` that asserts this. If you wish, you can open an issue for that.

Reply:
Validator issue created here: https://github.com/MaterializeInc/materialize/issues/11081